projects/driver_controls/src/stm32f0xx/ads1015.c

```diff
@@ -0,0 +1,183 @@
+                                     uint8_t lsb) {
+  uint8_t ads1015_setup_register[] = { reg, msb, lsb };
+  return i2c_write(storage->i2c_port, storage->i2c_addr, ads1015_setup_register,
+                   SIZEOF_ARRAY(ads1015_setup_register));
+}
+
+// Reads the register and stores the value in the given array.
+static StatusCode prv_read_register(I2CPort i2c_port, uint8_t i2c_addr, uint8_t reg,
+                                    uint8_t *rx_data, size_t rx_len) {
+  status_ok_or_return(i2c_write(i2c_port, i2c_addr, &reg, sizeof(reg)));
+
+  status_ok_or_return(i2c_read(i2c_port, i2c_addr, rx_data, rx_len));
+  return STATUS_CODE_OK;
+}
+
+// Switches to the given channel by writing to config register.
+static StatusCode prv_set_channel(Ads1015Storage *storage, Ads1015Channel channel) {
+  if (channel >= NUM_ADS1015_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  status_ok_or_return(prv_setup_register(storage, ADS1015_ADDRESS_POINTER_CONFIG,
+                                         ADS1015_CONFIG_REGISTER_MSB(channel),
+                                         ADS1015_CONFIG_REGISTER_LSB));
+  storage->current_channel = channel;
+  return STATUS_CODE_OK;
+}
+
+// This function is registered as the callback for ALRT/RDY Pin.
+// Reads and stores the conversion value in storage, and switches to the next enabled channel.
+// Also if there is a callback on a channel, it will be run here.
+static void prv_interrupt_handler(const GPIOAddress *address, void *context) {
+  Ads1015Storage *storage = context;
+  Ads1015Channel current_channel = storage->current_channel;
+  uint8_t channel_enable_bitset = storage->channel_enable_bitset;
+  uint8_t read_conv_register[2] = { 0, 0 };
+  if (channel_is_enabled(storage, current_channel)) {
+    prv_read_register(storage->i2c_port, storage->i2c_addr, ADS1015_ADDRESS_POINTER_CONV,
+                      read_conv_register, SIZEOF_ARRAY(read_conv_register));
+    // Following line puts the two read bytes into an int16.
+    // 4 least significant bits are not part of the result hence the bitshift.
+    storage->channel_readings[current_channel] =
+        ((read_conv_register[0] << 8) | read_conv_register[1]) >>
+        ADS1015_NUM_RESERVED_BITS_CONV_REG;
+    // Runs the users callback if not NULL.
+    if (storage->channel_callback[current_channel] != NULL) {
+      storage->channel_callback[current_channel](current_channel,
+                                                 storage->callback_context[current_channel]);
+    }
+  }
+
+  prv_mark_channel_enabled(current_channel, false, &storage->channel_enable_bitset_pending);
+  if (storage->channel_enable_bitset_pending == ADS1015_BITSET_EMPTY) {
+    // Reset the pending bitset once gone through a cycle of channel rotation.
+    storage->channel_enable_bitset_pending = channel_enable_bitset;
+  }
+  current_channel = __builtin_ffs(storage->channel_enable_bitset_pending) - 1;
```

Comment:
What happens if you have no channels active? Seems like you're relying on `current_channel` to underflow and `prv_set_channel` to handle that case for you.

Reply:
i could make that a function which would check for that and return a status code but the it_handler is void.
How do you guys usually handle errors in this case(in a void function)?

Reply:
You wouldn't - this is an interrupt handler, so you can't return anything.

It's more that you shouldn't start a new conversion if you have no channels active. That already happens because `current_channel` will underflow and not pass the channel check in `prv_set_channel`.